OCT is a standard pipeline for doing de novo transcriptomics. It is composed of three principle tools: an Assembly pipeline responsible for creating an assembly from reads, an Annotation pipeline responsible for gathering relevant information relating to your transcripts from multiple databases, and an Expression pipeline responsible for identifying deferentially expressed transcripts and genes in your reads. Each tool can be run individually, or all three can be executed as a single virtual tool, Full. 



## Input Output Overview
In order to run OCT, simply call it using...  
```
python pipeline.py [TOOL_SELECTOR] [ARGUMENTS]  
``` 
The [TOOL_SELECTOR] argument is what is used to specify what should be done by OCT. Accepted values are "assembly", "annotation", "expression", and "full". Using "assembly" will execute the assembly tool, using "annotation" will execute the annotation tool, and using "expression" will execute the expression tool. Using "full" will execute the assembly tool, followed by the expression tool and the annotation tool.

Regardless of what parameters are used or what tools are run, the following out_dir will be generated.

```
out_dir  
    annotation_files - contains files generated by the annotation tool
    assembly_files - contains files generated by the assembly tool
    expression_files - contains files generated by the expression tool
    log_files - Contains the stderr and stdout for everything executed by OCT.
    myassembly.fasta - The assembly generated by the assembly tool or copied in if provided
    run.log - A log specifying when specific jobs were started/finished.
```

For further information. consult the individual wiki pages for each of the tool options

[Full](Full_tool)  
[Assembly](Assembly_tool)  
[Annotation](Annotation_tool)  
[Expression](Expression_tool)  


